From 64e271844d00fa8531b5fcc3dbd722bd3160d472 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Fri, 21 Feb 2014 21:59:40 +0900
Subject: [PATCH 1/3] Support custom HTTP-version string (e.g., ICY)

---
 http_parser.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/http_parser.c b/http_parser.c
index df1b696..ccef696 100644
--- a/http_parser.c
+++ b/http_parser.c
@@ -238,6 +238,7 @@ enum state
   , s_start_req_or_res
   , s_res_or_resp_H
   , s_start_res
+  , s_res_proto_custom
   , s_res_H
   , s_res_HT
   , s_res_HTT
@@ -724,6 +725,10 @@ size_t http_parser_execute (http_parser *parser,
             break;
 
           default:
+            if(TOKEN(ch)) {
+              parser->state = s_res_proto_custom;
+              break;
+            }
             SET_ERRNO(HPE_INVALID_CONSTANT);
             goto error;
         }
@@ -732,6 +737,21 @@ size_t http_parser_execute (http_parser *parser,
         break;
       }
 
+      /* Custom HTTP-version string (e.g., ICY) */
+      case s_res_proto_custom:
+        if (ch == ' ') {
+          /* We assume this is HTTP/1.0 */
+          parser->http_major = 1;
+          parser->http_minor = 0;
+          parser->state = s_res_first_status_code;
+          break;
+        }
+        if (!TOKEN(ch)) {
+          SET_ERRNO(HPE_INVALID_CONSTANT);
+          goto error;
+        }
+        break;
+
       case s_res_H:
         STRICT_CHECK(ch != 'T');
         parser->state = s_res_HT;
-- 
1.9.2

