From 0230de8d757d45dad9ab9fa8cb84613b53eb80b6 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Fri, 21 Feb 2014 21:59:40 +0900
Subject: [PATCH 1/3] Support custom HTTP-version string (e.g., ICY)

---
 http_parser.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/http_parser.c b/http_parser.c
index cbe8a90..c0825b0 100644
--- a/http_parser.c
+++ b/http_parser.c
@@ -282,6 +282,7 @@ enum state
   , s_start_req_or_res
   , s_res_or_resp_H
   , s_start_res
+  , s_res_proto_custom
   , s_res_H
   , s_res_HT
   , s_res_HTT
@@ -761,6 +762,10 @@ reexecute:
             break;
 
           default:
+            if(TOKEN(ch)) {
+              parser->state = s_res_proto_custom;
+              break;
+            }
             SET_ERRNO(HPE_INVALID_CONSTANT);
             goto error;
         }
@@ -769,6 +774,21 @@ reexecute:
         break;
       }
 
+      /* Custom HTTP-version string (e.g., ICY) */
+      case s_res_proto_custom:
+        if (ch == ' ') {
+          /* We assume this is HTTP/1.0 */
+          parser->http_major = 1;
+          parser->http_minor = 0;
+          parser->state = s_res_first_status_code;
+          break;
+        }
+        if (!TOKEN(ch)) {
+          SET_ERRNO(HPE_INVALID_CONSTANT);
+          goto error;
+        }
+        break;
+
       case s_res_H:
         STRICT_CHECK(ch != 'T');
         UPDATE_STATE(s_res_HT);
-- 
2.1.4

