diff --git a/src/http-parser/http_parser.c b/src/http-parser/http_parser.c
index 9695525..98b7b20 100644
--- a/src/http-parser/http_parser.c
+++ b/src/http-parser/http_parser.c
@@ -1633,9 +1633,14 @@ size_t http_parser_execute (http_parser *parser,
 
         /* Exit, the rest of the connect is in a different protocol. */
         if (parser->upgrade) {
-          parser->state = NEW_MESSAGE();
-          CALLBACK_NOTIFY(message_complete);
-          return (p - data) + 1;
+          /* We want to use http_parser for tunneling connection
+             transparently */
+          /* Read body until EOF */
+          parser->state = s_body_identity_eof;
+          break;
+          /* parser->state = NEW_MESSAGE(); */
+          /* CALLBACK_NOTIFY(message_complete); */
+          /* return (p - data) + 1; */
         }
 
         if (parser->flags & F_SKIPBODY) {
